### PR TITLE
Filter routes by namespace or project labels

### PR DESCRIPTION
**Pull request by smarterclayton**
Allows `--namespace-labels=a=b` to filter the set of routes for a router to 
only namespaces that have the label (requires LIST namespaces) or 
`--project-labels=a=b` which shows projects you have access to with the given
label. You can also specify `--project-labels='*'` to show routes from all
projects you have access to.

Will allow sharding routes by arbitrary criteria - the next step is to make
routers be the ones who assign Host (and support multiple routers per route).

@derekwaynecarr for changes to project rest, @ramr @pweil- for route changes

### CODE REVIEW

**Comment by smarterclayton** [comment]
Builds on top of #4428

File: pkg/cmd/infra/router/template.go [hunk 1]
```diff
@@ -27,7 +27,10 @@ created by users and keeps a local router configuration up to date with those ch
 
 You may customize the router by providing your own --template and --reload scripts.
 
-You may restrict the set of routes exposed by using the --labels, --fields, or --namespace arguments.`
+You may restrict the set of routes exposed to a single project (with --namespace), projects your client has
+access to with a set of labels (--project-labels), namespaces matching a label (--namespace-labels), or all
+namespaces (no argument). You can limit the routes to those matching a --labels or --fields selector. Note
+that you must have a cluster-wide administrative role to view all namespaces.`
```
Comment (ramr) [inline pkg/cmd/infra/router/template.go]: Routing traffic to pods from different projects/namespaces will be an issue for now due to the networking isolation changes.

Comment (smarterclayton) [inline pkg/cmd/infra/router/template.go]: I'm not sure that network isolation will default to on.  However, there's
no assumption these scoped routers *must* run inside the pod network vs
outside.

On Sep 4, 2015, at 2:32 PM, Ram Ranganathan <notifications@github.com>
wrote:

In pkg/cmd/infra/router/template.go
<https://github.com/openshift/origin/pull/4462#discussion_r38781081>:

> @@ -27,7 +27,10 @@ created by users and keeps a local router configuration up to date with those ch
>
>  You may customize the router by providing your own --template and --reload scripts.
>
> -You may restrict the set of routes exposed by using the --labels, --fields, or --namespace arguments.`
> +You may restrict the set of routes exposed to a single project (with --namespace), projects your client has
> +access to with a set of labels (--project-labels), namespaces matching a label (--namespace-labels), or all
> +namespaces (no argument). You can limit the routes to those matching a --labels or --fields selector. Note
> +that you must have a cluster-wide administrative role to view all namespaces.`

Routing traffic to pods from different projects/namespaces will be an issue
for now due to the networking isolation changes.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4462/files#r38781081>.

**Comment by ramr** [comment]
LGTM  pending the gofmt changes  to ```./plugins/router/template/plugin_test.go```  and  
```./plugins/router/template/router.go``` which caused the tests to fail.

**Comment by smarterclayton** [comment]
[merge] on green

**Pull request merged**